src/net_processing.cpp

```diff
@@ -1723,6 +1730,16 @@ void PeerManagerImpl::BlockConnected(const std::shared_ptr<const CBlock>& pblock
             m_txrequest.ForgetTxHash(ptx->GetWitnessHash());
         }
     }
+
+    // In case the dynamic timeout was doubled once or more, reduce it slowly back to its default value
```

Comment:
@sipa Do you think that your proposed change should be implemented in this PR or can it be done in a followup? From my perspective, this PR seems to be strictly an improvement even if the stalling timeout backs off too aggressively.

Reply:
I haven't had time to look into this feedback closely yet - but I am planning to do that next week.

Reply:
I think this could be a good follow-up, but anyway, here are my thoughts.

>Decrease it when it didn't trigger due to being high enough.

I think this is a very good abstract policy.

>when the stalling detection triggers, and the stalling timer starts, but then the timeout is not reached.

This sounds more efficient at doing what it's supposed to do than what's implemented now.

>; e.g. we could set the new timeout to (old_timeout + actual_time_used) / 2.

Sounds like a good policy, but not going lower than 2 seconds probably. One could do some math modeling, but I don't think it's that helpful:
1) with random data, there is no ground truth — one would have to rely on the human sanity check of the inputs, which we already do verbally here;
2) could be tested against a couple laggy nodes too, comparing between different policies, but eh.

Reply:
> From my perspective, this PR seems to be strictly an improvement even if the stalling timeout backs off too aggressively.

I'd say that with its current approach, the PR improves behavior in isolated stalling situations where there are currently repeated disconnections of many/all of our peers without making any progress in getting blocks - but a previous stalling situation will not affect the behavior in future stalling situations because all memory of the previous stalling incident will be lost after a few blocks:

@sipa's suggestion would introduce a long-lasting memory of previous stalling situations
- I think that one downside of this approach is that the moving window algorithm should only lead to a stalling situation if one peer is significantly slower than the rest of the peers. We'd want to replace this first peer usually - giving it more time based on previous stalling situations is probably not beneficial, because if it was comparably fast to other peers, this would not have led to this peer being marked as a staller in the first place.
- The upside is in situations where the actual time to download a block for us is significantly larger than 2 seconds - we'd churn through multiple peers / timeout doublings again in every stalling situation right now until we reach the "correct" timeout, but wouldn't anymore with a longer memory.

> * Decrease it when it didn't trigger _due to being high enough_. And I think we have a way of measuring that: when the stalling detection triggers, and the stalling timer starts, but then the timeout is not reached. And by seeing how long it actually took for before the stalling state is resolved we can even do better than just applying a % drop; e.g. we could set the new timeout to (old_timeout + actual_time_used) / 2.

I think this would mean moving to the decrease of the stalling timeout to `ProcessMessages` (`NetMsgType::BLOCK`) where `m_stalling_since` is currently reset back to 0. At this point we haven't validated the block yet or connected it to the chain, wo we'd likely would need to at least make sure that we only decrease it after receiving the actual block that allows us to extend our chain (the peer might also have been sending us another block).